frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -0,0 +1,1452 @@
+	Feasibility(FeasibilityError),
+	/// An error in the miner (offchain) sub-system.
+	Miner(unsigned::MinerError),
+	/// An error in the on-chain fallback.
+	OnChainFallback(onchain::Error),
+	/// No fallback is configured. This is a special case.
+	NoFallbackConfigured,
+}
+
+impl From<onchain::Error> for ElectionError {
+	fn from(e: onchain::Error) -> Self {
+		ElectionError::OnChainFallback(e)
+	}
+}
+
+impl From<FeasibilityError> for ElectionError {
+	fn from(e: FeasibilityError) -> Self {
+		ElectionError::Feasibility(e)
+	}
+}
+
+impl From<unsigned::MinerError> for ElectionError {
+	fn from(e: unsigned::MinerError) -> Self {
+		ElectionError::Miner(e)
+	}
+}
+
+/// Errors that can happen in the feasibility check.
+#[derive(Debug, Eq, PartialEq)]
+pub enum FeasibilityError {
+	/// Wrong number of winners presented.
+	WrongWinnerCount,
+	/// The snapshot is not available.
+	///
+	/// Kinda defensive: The pallet should technically never attempt to do a feasibility check when
+	/// no snapshot is present.
+	SnapshotUnavailable,
+	/// Internal error from the election crate.
+	NposElection(sp_npos_elections::Error),
+	/// A vote is invalid.
+	InvalidVote,
+	/// A voter is invalid.
+	InvalidVoter,
+	/// A winner is invalid.
+	InvalidWinner,
+	/// The given score was invalid.
+	InvalidScore,
+	/// The provided round is incorrect.
+	InvalidRound,
+}
+
+impl From<sp_npos_elections::Error> for FeasibilityError {
+	fn from(e: sp_npos_elections::Error) -> Self {
+		FeasibilityError::NposElection(e)
+	}
+}
+
+pub use pallet::*;
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + SendTransactionTypes<Call<Self>> {
+		type Event: From<Event<Self>>
+			+ Into<<Self as frame_system::Config>::Event>
+			+ IsType<<Self as frame_system::Config>::Event>;
+
+		/// Currency type.
+		type Currency: ReservableCurrency<Self::AccountId> + Currency<Self::AccountId>;
+
+		/// Duration of the unsigned phase.
+		#[pallet::constant]
+		type UnsignedPhase: Get<Self::BlockNumber>;
+		/// Duration of the signed phase.
+		#[pallet::constant]
+		type SignedPhase: Get<Self::BlockNumber>;
+
+		/// The minimum amount of improvement to the solution score that defines a solution as
+		/// "better" (in any phase).
+		#[pallet::constant]
+		type SolutionImprovementThreshold: Get<Perbill>;
+
+		/// The priority of the unsigned transaction submitted in the unsigned-phase
+		type MinerTxPriority: Get<TransactionPriority>;
+		/// Maximum number of iteration of balancing that will be executed in the embedded miner of
+		/// the pallet.
+		type MinerMaxIterations: Get<u32>;
+		/// Maximum weight that the miner should consume.
+		///
+		/// The miner will ensure that the total weight of the unsigned solution will not exceed
+		/// this values, based on [`WeightInfo::submit_unsigned`].
+		type MinerMaxWeight: Get<Weight>;
```

Comment:
Any reason all gets shouldnt be exposed as constants?

Reply:
not really. This is a configuration to tell how much weight the miner embedded in this pallet should use. it is not an indication of how the maximum weight allowed by the chain. I think given that, it is not useful to anyone.